src/bun.js/bindings/JSPropertyIterator.zig

```diff
@@ -21,38 +14,58 @@ pub fn JSPropertyIterator(comptime options: JSPropertyIteratorOptions) type {
         len: usize = 0,
         i: u32 = 0,
         iter_i: u32 = 0,
-        impl: ?*anyopaque = null,
+        /// null if and only if `object` has no properties (i.e. `len == 0`)
+        impl: ?*JSPropertyIteratorImpl = null,
 
         globalObject: *JSC.JSGlobalObject,
-        object: *JSC.JSCell = undefined,
+        object: *JSC.JSObject,
+        // current property being yielded
         value: JSC.JSValue = .zero,
 
         pub fn getLongestPropertyName(this: *@This()) usize {
-            if (this.impl == null) return 0;
-            return Bun__JSPropertyIterator__getLongestPropertyName(this.impl, this.globalObject, this.object);
+            return if (this.impl) |iter|
+                iter.getLongestPropertyName(this.globalObject, this.object)
+            else
+                0;
         }
 
         pub fn deinit(this: *@This()) void {
-            if (this.impl) |impl| {
-                Bun__JSPropertyIterator__deinit(impl);
-            }
+            if (this.impl) |impl| impl.deinit();
             this.* = undefined;
         }
 
+        /// `object` should be a `JSC::JSObject`. Non-objects will be runtime converted.
         pub fn init(globalObject: *JSC.JSGlobalObject, object: JSC.JSValue) bun.JSError!@This() {
```

Comment:
```suggestion
        pub fn init(globalObject: *JSC.JSGlobalObject, object: *JSC.JSObject) bun.JSError!@This() {
```
This will allow us to avoid the exception. Most changes needed to make this work look like:
```diff
- if (!value.isObject()) {
+ const obj = value.getObject() orelse {
    global.throw(...);
}
```
}

Reply:
I'll make this change in a follow-up PR.